### PR TITLE
Add bloom filter

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: bloom/bloomfilter.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		Exist(hash.Hash256) bool
+		// Bytes returns the bytes of bloom filter
+		Bytes() []byte
+		// SetBytes copies the bytes into bloom filter
+		SetBytes([]byte) error
+	}
+)
+
+// NewBloomFilter returns a new bloom filter
+func NewBloomFilter(m, h int) (BloomFilter, error) {
```
Comment (zjshen14) [inline bloom/bloomfilter.go]: uint so that you don't need to check <0's case

File: bloom/bloomfilter.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	// BloomFilter interface
+	BloomFilter interface {
+		// Add 32-byte key into bloom filter
+		Add(hash.Hash256)
+		// Exist checks if a key is in bloom filter
+		Exist(hash.Hash256) bool
+		// Bytes returns the bytes of bloom filter
+		Bytes() []byte
+		// SetBytes copies the bytes into bloom filter
+		SetBytes([]byte) error
```
Comment (zjshen14) [inline bloom/bloomfilter.go]: We may not need SetBytes api given BloomFilterFromBytes exists

File: bloom/bloom2048b.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+type (
+	// bloom2048b implements a 2048-bit bloom filter
+	bloom2048b struct {
+		array   [256]byte
+		numHash int // number of hash function
+	}
+)
+
+// Add 32-byte key into bloom filter
+func (f *bloom2048b) Add(key hash.Hash256) {
```
Comment (zjshen14) [inline bloom/bloom2048b.go]: byte slice?

File: bloom/bloom2048b.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	h := hash.Hash256b(key[:])
+	// each 2-byte pair used as output of hash function
+	for i := 0; i < f.numHash; i++ {
+		f.setBit(h[2*i], h[2*i+1])
+	}
+}
+
+// Exist checks if a key is in bloom filter
+func (f *bloom2048b) Exist(key hash.Hash256) bool {
```
Comment (zjshen14) [inline bloom/bloom2048b.go]: ditto

File: bloom/bloomfilter.go (hunk 1 above)
Comment (dustinxie) [reply bloom/bloomfilter.go]: fixed in new push

File: bloom/bloomfilter.go (hunk 2 above)
Comment (dustinxie) [reply bloom/bloomfilter.go]: fixed in new push

File: bloom/bloom2048b.go (hunk 3 above)
Comment (dustinxie) [reply bloom/bloom2048b.go]: fixed in new push

**Pull request merged**